packages/datadog-instrumentations/src/mariadb.js

```diff
@@ -0,0 +1,83 @@
+'use strict'
+
+const {
+  channel,
+  addHook,
+  AsyncResource
+} = require('./helpers/instrument')
+
+const shimmer = require('../../datadog-shimmer')
+
+const startCh = channel('apm:mariadb:query:start');
+const finishCh = channel('apm:mariadb:query:finish');
+const errorCh = channel('apm:mariadb:query:error');
+
+addHook(
+  {
+    name: 'mariadb',
+    file: 'lib/connection-promise.js',
+    versions: ['>=3'],
+  },
+  (ConnectionPromise) => {
+    shimmer.wrap(ConnectionPromise, '_QUERY_CMD', (queryCmd) => {
+      return function (_conn, sql, _values) {
+        if (!startCh.hasSubscribers) {
+          return queryCmd.apply(this, arguments);
+        }
+
+        const asyncResource = new AsyncResource('bound-anonymous-fn');
+
+        return asyncResource.runInAsyncScope(() => {
+          startCh.publish({ sql });
+
+          try {
+            const promise = queryCmd.apply(this, arguments);
+
+            promise.then(
+              asyncResource.bind(() => finishCh.publish()),
+              asyncResource.bind((e) => {
+                errorCh.publish(e);
+                finishCh.publish();
+              })
+            );
+
+            return promise;
+          } catch (e) {
+            errorCh.publish(e);
+            finishCh.publish();
+            throw e;
+          }
+        });
+      };
+    });
+
+    shimmer.wrap(ConnectionPromise.prototype, 'queryStream', (queryStream) => {
+      return function (sql, _values) {
+        if (!startCh.hasSubscribers) {
+          return queryStream.apply(this, arguments);
+        }
+
+        startCh.publish({ sql });
+
+        try {
```

Comment:
This block should be wrapped with an `AsyncResource` to ensure that the store change from the subscriber doesn't leak beyond this function.